### PR TITLE
Refactoring of the KQL grammar.

### PR DESCRIPTION
**Pull request by afoucret**
KQL grammar refactoring to make it easier to implement the AST query builder.

### CODE REVIEW

File: x-pack/plugin/kql/src/test/resources/supported-queries [hunk 1]
```diff
@@ -67,6 +67,8 @@ foo_field:foo OR foo_field:foo bar AND foo bar
 foo_field:foo AND (foo_field:foo bar OR foo bar)
 foo_field:foo AND (foo_field:foo bar OR foo bar)
 foo_field:foo OR (foo_field:foo bar OR foo bar)
+foo AND
```
Comment (afoucret) [inline x-pack/plugin/kql/src/test/resources/supported-queries:70]: ℹ️ KQL support queries with a trailing AND operator and just ignore it.

The grammar has been updated to support it and these queries have been moved from `unsupported-queries` to `supported-queries`

File: x-pack/plugin/kql/src/main/antlr/KqlBase.g4 [hunk 2]
```diff
@@ -26,70 +26,66 @@ topLevelQuery
     ;
 
 query
-    : query (AND | OR) query          #booleanQuery
-    | NOT subQuery=simpleQuery        #notQuery
-    | simpleQuery                     #defaultQuery
+    : <assoc=right> query operator=(AND | OR) query?    #booleanQuery
```
Comment (afoucret) [inline x-pack/plugin/kql/src/main/antlr/KqlBase.g4:29]: ℹ️ Adding `<assoc=right>` changes the operator precedence and was required to match KQL behavior.
There will be a specific test for it as shown in https://github.com/elastic/elasticsearch/pull/115632:

```java
// <QueryA> AND <QueryB> OR <QueryC> is equivalent to <QueryA> AND (<QueryB> OR <QueryC>)
assertThat(
    parser.parseKqlQuery(Strings.format("%s AND %s OR %s", queryA, queryB, queryC), searchExecutionContext),
    equalTo(parser.parseKqlQuery(Strings.format("%s AND (%s OR %s)", queryA, queryB, queryC), searchExecutionContext))
);

// <QueryA> OR <QueryB> AND <QueryC> is equivalent to <QueryA> OR (<QueryB> AND <QueryC>)
assertThat(
    parser.parseKqlQuery(Strings.format("%s OR %s AND %s", queryA, queryB, queryC), searchExecutionContext),
    equalTo(parser.parseKqlQuery(Strings.format("%s OR (%s AND %s)", queryA, queryB, queryC), searchExecutionContext))
);
```

File: x-pack/plugin/kql/src/test/resources/unsupported-queries [hunk 3]
```diff
@@ -18,9 +18,7 @@ foo_field: foo bar"
 
 
 // Invalid boolean queries
-foo AND
```
Comment (carlosdelest) [inline x-pack/plugin/kql/src/test/resources/unsupported-queries:21]: `AND foo` is, surprisingly, supported

image

**Review by carlosdelest — APPROVED** [review]
LGTM - one question about wildcards

File: x-pack/plugin/kql/src/main/antlr/KqlBase.g4 [hunk 4]
```diff
@@ -26,70 +26,66 @@ topLevelQuery
...
-    ;
-
-expression
-    : fieldTermQuery
-    | fieldRangeQuery
+    | matchAllQuery
+    | existsQuery
+    | rangeQuery
+    | fieldQuery
+    | fieldLessQuery
```
Comment (afoucret) [inline x-pack/plugin/kql/src/main/antlr/KqlBase.g4:41]: `fieldTermQuery` has been split into several queries (`matchAllQuery`, `existQuery`, `fieldQuery`, `fieldLessQuery`) so a lot of conditional logic can be removed from the code.

File: x-pack/plugin/kql/src/test/resources/unsupported-queries (hunk 3 above)
Comment (afoucret) [reply x-pack/plugin/kql/src/test/resources/unsupported-queries:21]: Updated the grammar so we can run these queries.

File: x-pack/plugin/kql/src/main/antlr/KqlBase.g4 [hunk 5]
```diff
@@ -98,24 +94,24 @@ OR: 'or';
...
 LEFT_CURLY_BRACKET: '{';
 RIGHT_CURLY_BRACKET: '}';
 
-UNQUOTED_LITERAL: WILDCARD* UNQUOTED_LITERAL_CHAR+ WILDCARD*;
+UNQUOTED_LITERAL: WILDCARD* UNQUOTED_LITERAL_CHAR+ WILDCARD* | WILDCARD_CHAR WILDCARD+;
 
 QUOTED_STRING: '"'QUOTED_CHAR*'"';
 
-WILDCARD: WILDCARD_CHAR+;
+WILDCARD: WILDCARD_CHAR;
```
[the remark this answers is not in the document]
Comment (afoucret) [reply x-pack/plugin/kql/src/main/antlr/KqlBase.g4:113]: `WILDCARD` is a token which is exposed by the lexer while `WILDCARD_CHAR` is a fragment that is not exposed.

`WILDCARD_CHAR` is used to build other fragments such as `UNQUOTED_LITERAL_CHAR`, which can not be done with a token and require a fragment.

**Pull request merged**